shell/platform/common/client_wrapper/core_implementations.cc

```diff
@@ -36,17 +36,31 @@ void ForwardToHandler(FlutterDesktopMessengerRef messenger,
                       const FlutterDesktopMessage* message,
                       void* user_data) {
   auto* response_handle = message->response_handle;
-  BinaryReply reply_handler = [messenger, response_handle](
+  auto messenger_ptr = std::shared_ptr<FlutterDesktopMessenger>(
+      FlutterDesktopMessengerAddRef(messenger),
+      &FlutterDesktopMessengerRelease);
+  BinaryReply reply_handler = [messenger_ptr, response_handle](
                                   const uint8_t* reply,
                                   size_t reply_size) mutable {
+    // Note: This can be called on any thread.
+    auto lock = std::unique_ptr<FlutterDesktopMessenger,
+                                decltype(&FlutterDesktopMessengerUnlock)>(
+        FlutterDesktopMessengerLock(messenger_ptr.get()),
+        &FlutterDesktopMessengerUnlock);
+    if (!FlutterDesktopMessengerIsAvailable(messenger_ptr.get())) {
+      std::cerr << "Error: Responding to platform channel after the engine has"
+                   "been deleted."
```

Comment:
Is this really an error? If the engine is gone, then this is a harmless no-op, and it doesn't seem like there's anything wrong with the calling code (unlike duplicate reply below, which suggests a serious logic error in the plugin).

I'm trying to figure out what the actionable change someone should make here if they were to see this log message. It seems like it would be to add some annoying cross-thread engine shutdown tracking for the sole purpose of avoiding sending the reply... which is what this code is already doing internally, but logging an error for. So plugin authors would need to add complexity for the sole purpose of avoiding this error *message*.

Is there a case I'm not seeing where this actually indicates a meaningful error by the plugin developer? If not, this should be comment explaining why we're dropping it, rather than a log.

Reply:
It's an error for a plugin developer if they've set up their test harness wrong.  They might have an ASSERT_EQ somewhere that isn't actually triggering because they are shutting down the engine before channels respond.

It's an error for add-to-app developers if they are creating and destroying engines and some important piece of data was supposed to be delivered to the host app before shutting down a flutter engine.

Since the scenario could be flakey I think it's best not to be silent about it.  Also it could represent a huge drop of data if the environment is not setup correctly.

Reply:
> It's an error for a plugin developer if they've set up their test harness wrong. They might have an ASSERT_EQ somewhere that isn't actually triggering because they are shutting down the engine before channels respond.

I'm not sure I follow this one. How would you write a plugin test that shuts down the engine before the test completes?

> It's an error for add-to-app developers if they are creating and destroying engines and some important piece of data was supposed to be delivered to the host app before shutting down a flutter engine.

What would an add-to-app developer do if they saw this message? They don't have any control over, or visibility into, threads created by plugins, so I don't see how they could possibly know when all pending replies had come in.

> Since the scenario could be flakey I think it's best not to be silent about it.

I'm trying to figure out how to reconcile this with https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#only-log-actionable-messages-to-the-console. The only "fix" at the plugin level is to add a lot of complexity (an extra layer of tracking and locking) in *every single plugin using this functionality*, despite the fact that in the common case this is not, in fact, an error.

If someone had a simple plugin that, say, read something from the registry on a background thread and returned it, every single time an app using that plugin quits, there's a potential race that could log this error message. What is the actionable step for the plugin author? If the only answer is "add main-thread-locking engine shutdown tracking for the sole purpose of avoiding this error message when nothing is actually wrong", that's a really bad API experience. It's not just unnecessary annoying complexity for the plugin author, it will slow down shutdown (increasingly with every plugin using the API).

Reply:
There are plugins that start their own engines.  It isn't a pattern we condone, but I don't think we'll be able to block it.

The add-to-app developer would have to synchronize on the reply before shutting down the engine instead of keeping the racy situation they had.

For the hypothetical plugin that reads asynchronously from a registry, it would be the app developers decision wether replying on a dead channel was acceptable or not.  Imagine that it's a mac app that uploads large videos to the web.  It opens a floating window that has its own engine for displaying a progress bar and uploading.  When the upload is done it does a registry check to make sure the hour spent uploading the video was ingested correctly.  Then the window and the engine are killed.

The unit test added to the PR is an example where if written poorly could have resulted in this sort of flake.

Is the thought that printing this message does more harm than good?  That the false positive log is more harmful than being silent for a true positive?

Reply:
Or I think you are more arguing the existence of the true positive, right?  That this situation is not erroneous in any context?

Reply:
> it would be the app developers decision wether replying on a dead channel was acceptable or not

And if they decide it is, which I would expect to be overwhelmingly the common case, what would they do to prevent every user of their app from occasionally seeing a scary-sounding error message and reporting it to them as a bug? What is the actionable step for the client of my hypothetical plugin, getting this error report from an end user?

> Imagine that it's a mac app that uploads large videos to the web. It opens a floating window that has its own engine for displaying a progress bar and uploading. When the upload is done it does a registry check to make sure the hour spent uploading the video was ingested correctly. Then the window and the engine are killed.

If someone explicitly wants to wait for some async task to complete before shutting something down, they need to wait for it, sure. That's not specific to plugins though. Implement exactly your scenario, but with the registry logic in Dart, and you have the same problem. Should Dart VM shutdown log an error to console for every outstanding `Future`? If not, why is this different?

> Is the thought that printing this message does more harm than good? That the false positive log is more harmful than being silent for a true positive?

I think it does more harm than good, and that it violates an explicit Flutter policy about logging because it will, without a doubt, log error messages in cases that are not errors and that are therefore not actionable by any reasonable means I can see.

> Or I think you are more arguing the existence of the true positive, right? That this situation is not erroneous in any context?

I think the cases that you are describing are possible, but a) that they are significantly rarer than false positives, and b) that the errors are happening at a higher layer, and that at *this* layer we should not consider it an error.

Reply:
Okay, speaking from experience, the engine silently dropping messages has caused problems and hid race conditions in the past.  I may be overly sensitive as a result.  Replaced log with a comment.